p2pclient/p2pclient.go

```diff
@@ -0,0 +1,125 @@
+package p2pclient
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	ggio "github.com/gogo/protobuf/io"
+	logging "github.com/ipfs/go-log"
+	pb "github.com/libp2p/go-libp2p-daemon/pb"
+	multiaddr "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+var log = logging.Logger("p2pclient")
+
+// MessageSizeMax is cribbed from github.com/libp2p/go-libp2p-net
+const MessageSizeMax = 1 << 22 // 4 MB
+
+// Client is the struct that manages a connection to a libp2p daemon.
+type Client struct {
+	control manet.Conn
+
+	mhandlers sync.Mutex
+	handlers  map[string]io.Closer
+}
+
+// NewClient creates a new libp2p daemon client, connecting to a daemon
+// listening at the provided multiaddr.Multiaddr
+func NewClient(addr multiaddr.Multiaddr) (*Client, error) {
+	conn, err := manet.Dial(addr)
+	if err != nil {
+		return nil, err
+	}
+	cli := &Client{control: conn}
+	return cli, nil
+}
+
+// Close shuts down the control connection and all active handlers. Users are
+// responsible for closing the streams they've opened with the daemon.
+func (c *Client) Close() error {
+	c.mhandlers.Lock()
+	defer c.mhandlers.Unlock()
+
+	if err := c.control.Close(); err != nil {
+		return err
+	}
+
+	for proto, listener := range c.handlers {
+		if err := listener.Close(); err != nil {
+			return fmt.Errorf("closing listener for %s: %s", proto, err)
+		}
+	}
+
+	return nil
+}
+
+// Identify queries the daemon for its peer ID and listen addresses.
+func (c *Client) Identify() ([]byte, []multiaddr.Multiaddr, error) {
```

Comment:
This needs a lock in the control socket to support concurrent calls.
Alternatively, this could use a fresh unix socket for each request (that's what the Gerbil client does).